### PR TITLE
[ENH] Generic csv enh

### PR DESCRIPTION
**Pull request by navtej**
Apologies for a consolidated change set
Here are all the changes 

1.  Multiple values for columns.
    Frequently we come across feeds which can be easily handled with generic csv parser, however the parser often fails because of mixed content in some column eg FQDN/IP, IP/CIDR. This change allows one column to specify multiple possible values using '|' char eg `source.ip|source.network`
2.  Fix for weird timestamp values of length 12 and 13
    I came across some prop device feeds where the timestamp had addition two or three 000.  An attempted fix for such cases is implemented.
3.  data type for columns
    A possibility for data type conversion, similar to timestamp conversion. Currently only json data type is implemented.
eg `"data_type": "{\"extra.tags\":\"json\"}"` alongwith `{"columns": [ "source.ip", "source.url", "extra.tags"]`
4.  fix for empty url
         if the source omitted the value for .url column, existing code added http:// to it and proceeded to add the value which would fail. A check is added to handle this case.
5.  fix for regex test case
 although the test is focused on asn regex match but the testcase example event check never included the asn value. Fixes the omission. 
6.  .strip() values before adding to accommodate  white space between csv values
     feeds sometimes contain a space char between values separated with some delimiter. The IP values with leading/trailing space fail the harmonization checks, unless stripped.

**Updated:  23rd Aug**
Add ability for csv parser to skip/process line based on if the line contains or not contains some value.

### CODE REVIEW

File: intelmq/bots/parsers/generic/parser_csv.py [hunk 1]
```diff
@@ -66,30 +71,61 @@ def parse_line(self, row, report):
...
-                    extra[key[6:]] = value
-            else:
-                event.add(key, value)
+
+                if key in self.data_type:
+                    value = DATA_CONVERSIONS[self.data_type[key]](value)
+
+                if key in ["time.source", "time.destination"]:
+                    if self.time_format == 'timestamp':
+                        if len(value) == 12:
```
Comment (wagner-certat) [inline intelmq/bots/parsers/generic/parser_csv.py]: I'd prefer a new time conversion 'epoch' for this kind of format. And tests.

File: intelmq/bots/parsers/generic/parser_csv.py [hunk 2]
```diff
@@ -66,30 +71,61 @@ def parse_line(self, row, report):
...
+                regex = self.column_regex_search.get(key, None)
+                if regex:
+                    search = re.search(regex, value)
+                    if search:
+                        value = search.group(0)
+                    else:
+                        value = None
+
+                if key in ["__IGNORE__", ""]:
+                    stop_processing = True
```
Comment (wagner-certat) [inline intelmq/bots/parsers/generic/parser_csv.py]: See navtej#1

**Comment by navtej** [comment]
Sorry for the delay on this, i will work on it this week.

**Comment by wagner-certat** [comment]
@navtej Any updates in sight? I'd like to include these additions!

File: docs/Bots.md [hunk 3]
```diff
@@ -364,14 +364,30 @@ Lines starting with `'#'` will be ignored. Headers won't be interpreted.
...
+    Some use cases - 
+        - mixed data set, e.g. URL/FQDN/IP/NETMASK  `"columns": "source.url|source.fqdn|source.ip|source.network"`
+        - parse a value and ignore if it fails  `"columns": "source.url|__IGNORE__"`
  * `"column_regex_search"`: Optional. A dictionary mapping field names (as given per the columns parameter) to regular expression. The field is evaulated using `re.search`. Eg. to get the ASN out of `AS1234` use: `{"source.asn": "[0-9]*"}`.
  * `"default_url_protocol"`: For URLs you can give a defaut protocol which will be pretended to the data.
  * `"delimiter"`: separation character of the CSV, e.g. `","`
  * `"skip_header"`: Boolean, skip the first line of the file, optional. Lines starting with `#` will be skipped additionally, make sure you do not skip more lines than needed!
- * `time_format`: Optional. If `"timestamp"` or `"windows_nt"` the time will be converted first. With the default `null` fuzzy time parsing will be used.
+ * `time_format`: Optional. If `"timestamp"`, `"windows_nt"` or `"epoc"` the time will be converted first. With the default `null` fuzzy time parsing will be used.
```
Comment (wagner-certat) [inline docs/Bots.md]: Did you mean `epoch` instead of `epoc`?

Comment (navtej) [inline docs/Bots.md]: yes. I will fix it. I need fix the formatting as well.

File: intelmq/lib/harmonization.py [hunk 4]
```diff
@@ -230,6 +230,21 @@ def __parse_utc_isoformat(value):
         return value
 
     @staticmethod
+    def from_epoch_millis(tstamp, tzone='UTC'):
+        """
+        Returns ISO formatted datetime from given epoch timestamp with milliseconds.
+        It ignores the milliseconds, converts it into normal timestamp and processes it.
+        """
+        bytecount = len(str(tstamp))
+        int_tstamp = tstamp if isinstance(tstamp, int) else int(tstamp)
```
Comment (wagner-certat) [inline intelmq/lib/harmonization.py]: What is that for? If you call `int()` on a int object, the same object is returned.

**Comment by wagner-certat** [comment]
merge with develop and some fixes including a rename of `text` to `filter` are in navtej/intelmq#3

**Comment by wagner-certat** [comment]
Merged with my additions/changes

**Pull request merged**